### ISSUE TITLE
Get the ElementText rectangle or bounding box when the layout calculation is complete

### ISSUE THREAD

**Issue by Shen9ke** [mikke89/RmlUi#624]
Hi,
I am trying to develop a rich text editor with RmlUI, but the existing WidgetTextInput element cannot edit and selection elements with child nodes, but only plain text, so I want to try to implement Selection and Range methods based on browser principles. 
In order to achieve cross-text node and element node selection, I had to recalculate the text node layout to get the line feed information, which was too difficult. It would be nice to have an API in place to retrieve the calculated TextLayout (which contains the Rectangle List information) at rendering time, so as to  avoid  compute the text layout repeatedly. 
Or  have any better suggestions for implementing the Selection and Range methods?

**Comment by mikke89** [comment]
Hey there!

It's an interesting challenge. I guess we at least could start with exposing the [`lines` of text elements](https://github.com/mikke89/RmlUi/blob/e02ff487a5d61bcfd795cfbb1bc2e41911ce2892/Include/RmlUi/Core/ElementText.h#L108), from those you should be able to deduce the area of each line at least. This would be considered fairly low-level, I don't want to guarantee API stability here, because we want room to refactor. We could consider adding a more stable API to get the surrounding box of each line, or something like that.

Note that, we don't really store the position of each letter, which might be what you're after. You might want to take a look at how the WidgetTextInput does text selection, and generalize some of that so that it works for ElementText. The main challenge though is that for normal text layout, the layout engine does all the layouting, and that has a fairly complex state, considering all of the CSS properties and its layout intricacies, as I'm sure you can imagine. And we don't really want to add more state for text nodes that don't need selection. I would be a lot more concerned over adding more state, than the extra computation of recalculating some lines, if it comes to that.

You might also want to consider implementing a rich text editor using a custom element instead. That should give you a lot more control over everything, and probably wouldn't need any changes to the library. You lose the ability to take advantage of all the CSS layout features the library supports, if you need those. I guess it depends a bit on how much you want to dive into the library to add the necessary changes, and discuss the design around that in detail. Versus just get something up and running without too much effort and iterate on that, more independently of the library.

**Comment by Shen9ke** [comment]
Yeah, [lines](https://github.com/mikke89/RmlUi/blob/e02ff487a5d61bcfd795cfbb1bc2e41911ce2892/Include/RmlUi/Core/ElementText.h#L108) might work better, but as you pointed out, I noticed it is a private variable in the `ElementText `class. I don't want to use that through a awkward way. I prefer to rely on the standard RmlUi library and make no changes.
I noticed that the `WidgetTextInput `had used the FormatText method to handle the computation of lines and generate selection geometry. What confused me was why we were not directly using the layout that is finished by the layout engine. For the RmlUi layout stream, does it mean we need to layout text elements twice? Correct me if I'm wrong.
So it's going to be very contradictory. On one hand, I really agree with your view that it's bad to have extra computation for recalculating and we don't really want to change RmlUi itself, but on the other hand, we don't have a ready-made API to achieve this.

Custom elements are a good idea, and I will consider using them in the future to implement some extended functions for the rich text editor, such as mathematical formulas. However, I don't want to use custom elements here. As you mentioned, custom elements seem to lose the ability to leverage all the CSS layout features supported by the library. Moreover, the workload doesn't seem to be reduced. We prefer to utilize RmlUi's original features and methods to implement basic functionalities as much as possible.

Actually, my ideal solution would be to execute `ElementUtilities::GetBoundingBox(rect, textElement_node, BoxArea::Auto)` and obtain the rectangular information of `textElement_node,` so that I can use `textElement_node->GenerateLine(...)` to calculate again line breaks by limiting the boundaries of the text field. Unfortunately, it seems that `GetBoundingBox` can only retrieve the `BoundingBox `of` textElement_node`'s parent, which is not what I want. This is because the parent might have many non-text children, which can affect the calculation of line break layouts. As a result, I can't get the desired outcome.

Therefore, in summary, I still need to recalculate the virtual layout of the text myself to obtain the relevant information. What I can do is to consider and handle as many scenarios as possible to ensure a normal text layout.

**Comment by mikke89** [comment]
Hello, I am looking over some older issues. There hasn't been much activity here in a while, and since I don't see anything immediately actionable here, I am closing this one. We might want to consider this application (rich text editor) when working on #661,
